packages/SwingSet/src/kernel/state/kernelKeeper.js

```diff
@@ -866,14 +880,75 @@ export default function makeKernelKeeper(kernelStorage, kernelSlog) {
     kvStore.set(`${kernelSlot}.data.slots`, capdata.slots.join(','));
   }
 
-  function cleanupAfterTerminatedVat(vatID) {
+  function removeVat(vatID) {
+    // Delete primary swingstore records for this vat, in preparation
+    // for (slow) deletion. After this, swingstore exports will omit
+    // this vat. This is called from the kernel's terminateVat, which
+    // initiates (but does not complete) deletion.
+    snapStore.stopUsingLastSnapshot(vatID);
+    transcriptStore.stopUsingTranscript(vatID);
+  }
+
+  /**
+   * Perform some cleanup work for a specific (terminated but not
+   * fully-deleted) vat, possibly limited by a budget. Returns 'done'
+   * (where false means "please call me again", and true means "you
+   * can delete the vatID now"), and a count of how much work was done
+   * (so the runPolicy can decide when to stop).
+   *
+   * @param {string} vatID
+   * @param {number} [budget]
+   * @returns {{ done: boolean, cleanups: number }}
+   *
+   */
+  function cleanupAfterTerminatedVat(vatID, budget = undefined) {
     insistVatID(vatID);
+    let cleanups = 0;
+    const work = {
+      exports: 0,
+      imports: 0,
+      kv: 0,
+      snapshots: 0,
+      transcripts: 0,
+    };
+    let spend = _did => false; // returns "stop now"
+    if (budget !== undefined) {
+      assert.typeof(budget, 'number');
+      spend = (did = 1) => {
+        assert(budget !== undefined); // hush TSC
+        cleanups += did;
+        budget -= did;
+        return budget <= 0;
+      };
+    }
+    const logWork = () => {
+      const w = {};
+      for (const prop of Object.getOwnPropertyNames(work)) {
+        if (work[prop] !== 0) {
+          w[prop] = work[prop];
+        }
+      }
+      kernelSlog?.write({ type: 'vat-cleanup', vatID, work: w });
```

Comment:
Suggested simplification for filtering out zero-valued properties:
```suggestion
      const nonZeroWork = objectMetaMap(work, desc => desc.value ? desc : undefined);
      kernelSlog?.write({ type: 'vat-cleanup', vatID, work: nonZeroWork });
```
where `objectMetaMap` comes from `@endo/common/object-meta-map.js` via `@agoric/internal`.

Reply:
hm, looks nice, but it seems that `objectMetaMap` is not currently present in `@agoric/internal` (only `objectMap`).. you ok with me adding that?

Reply:
Definitely!